Surprising I2C `TimeOut` [ESP32-C3]

Reporting this so we can argue whether it's intended behaviour or not.

I'm taking my first steps with ESP32-C3 and I2C. Here's a log:

```
[TRACE] Writing: 0x7fff <- 1 bytes
[TRACE] I2C written: 0x7fff <- [0x09]
[TRACE] Writing: 0x0000 <- 252 bytes
[TRACE] I2C written: 0x0000 <- [0xe0, 0x00, 0x03, ...]
[TRACE] Writing: 0x00fc <- 252 bytes
[ERROR] I2C write failed: TimeOut
```

Here, I'm starting to loop each max. slice (that fits into the I2C FIFO) of a larger (32k) slice. I am somewhat surprised to receive a timeout, since the I2C is created with:

```rust
    let i2c = I2C::new(
        peripherals.I2C0,
        io.pins.gpio4,  // SDA
        io.pins.gpio5,  // SCL
        400.kHz(),
        &clocks,
        None
    );
```

There is the `new_with_timeout` variant that I'll be soon trying out. Here comes my first question:

- Should `::new` mean that there are no timeouts?  I would expect either that, or a "default" timeout, but the [documentation](https://docs.rs/esp-hal/latest/esp_hal/i2c/struct.I2C.html) is really scarce on `i2c`.
 
**Question #2**

- What is the unit of `timeout` parameter here: https://github.com/esp-rs/esp-hal/blob/main/esp-hal/src/i2c.rs#L380

   I'm kind of fine looking at these things within the `esp-hal` repo (have it checked out, locally), but this is going elsewhere, without a hint of information what the `u32` is. Likely, it'll be milliseconds?  But what if I want to *not* have timeouts?  Can I use `0`?

   >A: `0` times out immediately. `10` gives the default (same as with `::new`) behaviour, `100` and `1000` panic with:
   >
   >```
   >[ERROR] "panicked at /home/ubuntu/.cargo/registry/src/index.crates.io-6f17d22bba15001f/esp-hal-0.19.0/src/i2c.rs:1422:57:\ncalled `Result::unwrap()` on an `Err` value: TryFromIntError(())"
   >``` 

**Question #3**

I noticed, going through the code that internal SDA & SCL pull-ups are enabled for the bus. I have external ones - how should I not have them within the chip?

The documentation states:

>"[...] the usual setting such as frequency, timeout, and SDA/SCL pins can easily be configured."

I hope it's that way. Only, I don't know where to look for how easy it would be. Let's fix this!

---

**Edit:**

When coming to the `TimeOut`, the bus has already taken care of 30..40 1-byte reads and writes. It is the **cumulative size of transferred data** that seems to trigger the timeout. If I vary the size of the written chunks (tried 254, 100 and 50), I get more chunks through, but roughly the same overall data (400..500 bytes).

This is weird. I would expect a timeout to apply separately for each I2C transaction.

Same behaviour with both 100kHz and 400kHz bus speeds.

Versions:

- `esp-hal` 0.19.0 with features `"defmt", "esp32c3", "embedded-hal"`

--- a/esp-hal/CHANGELOG.md
+++ b/esp-hal/CHANGELOG.md
@@ -81,6 +81,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Fixed ESP32-S2 systimer interrupts (#1979)
 - Software interrupt 3 is no longer available when it is required by `esp-hal-embassy`. (#2011)
 - ESP32: Fixed async RSA (#2002)
+- Fix i2c embedded-hal transaction (#2028)
 
 ### Removed
 

--- a/esp-hal/CHANGELOG.md
+++ b/esp-hal/CHANGELOG.md
@@ -38,6 +38,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - A build issue when including doc comment prelude (#2040)
+- Fix i2c embedded-hal transaction (#2028)
 
 ## [0.20.0] - 2024-08-29
 
@@ -81,7 +82,6 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 - Fixed ESP32-S2 systimer interrupts (#1979)
 - Software interrupt 3 is no longer available when it is required by `esp-hal-embassy`. (#2011)
 - ESP32: Fixed async RSA (#2002)
-- Fix i2c embedded-hal transaction (#2028)
 
 ### Removed
 

--- a/esp-hal/CHANGELOG.md
+++ b/esp-hal/CHANGELOG.md
@@ -23,6 +23,7 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 
 - Fixed an issue with DMA transfers potentially not waking up the correct async task (#2065)
 - Fixed an issue with LCD_CAM i8080 where it would send double the clocks in 16bit mode (#2085)
+- Fix i2c embedded-hal transaction (#2028)
 
 ### Removed
 - Removed `digest::Digest` implementation from SHA (#2049)
@@ -38,7 +39,6 @@ and this project adheres to [Semantic Versioning](https://semver.org/spec/v2.0.0
 ### Fixed
 
 - A build issue when including doc comment prelude (#2040)
-- Fix i2c embedded-hal transaction (#2028)
 
 ## [0.20.0] - 2024-08-29
 

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -281,7 +281,7 @@ where
                         address,
                         bytes,
                         last_op != Op::Write,
-                        op_iter.peek().is_none(),
+                        next_op == Op::None,
                         cmd_iterator,
                     )?;
                     last_op = Op::Write;
@@ -295,7 +295,7 @@ where
                         address,
                         buffer,
                         last_op != Op::Read,
-                        op_iter.peek().is_none(),
+                        next_op == Op::None,
                         next_op == Op::Read,
                         cmd_iterator,
                     )?;
@@ -740,6 +740,14 @@ mod asynch {
             Ok(())
         }
 
+        /// Executes an async I2C write operation.
+        /// - `addr` is the address of the slave device.
+        /// - `bytes` is the data two be sent.
+        /// - `start` indicates whether the operation should start by a START
+        ///   condition and sending the address.
+        /// - `stop` indicates whether the operation should end with a STOP
+        ///   condition.
+        /// - `cmd_iterator` is an iterator over the command registers.
         async fn write_operation<'a, I>(
             &self,
             address: u8,
@@ -772,6 +780,16 @@ mod asynch {
             Ok(())
         }
 
+        /// Executes an async I2C read operation.
+        /// - `addr` is the address of the slave device.
+        /// - `buffer` is the buffer to store the read data.
+        /// - `start` indicates whether the operation should start by a START
+        ///   condition and sending the address.
+        /// - `stop` indicates whether the operation should end with a STOP
+        ///   condition.
+        /// - `will_continue` indicates whether there is another read operation
+        ///   following this one and we should not nack the last byte.
+        /// - `cmd_iterator` is an iterator over the command registers.
         async fn read_operation<'a, I>(
             &self,
             address: u8,
@@ -929,7 +947,7 @@ mod asynch {
                             address,
                             bytes,
                             last_op != Op::Write,
-                            op_iter.peek().is_none(),
+                            next_op == Op::None,
                             cmd_iterator,
                         )
                         .await?;
@@ -943,7 +961,7 @@ mod asynch {
                             address,
                             buffer,
                             last_op != Op::Read,
-                            op_iter.peek().is_none(),
+                            next_op == Op::None,
                             next_op == Op::Read,
                             cmd_iterator,
                         )
@@ -1463,6 +1481,11 @@ pub trait Instance: crate::private::Sealed {
     }
 
     /// Configures the I2C peripheral for a write operation.
+    /// - `addr` is the address of the slave device.
+    /// - `bytes` is the data two be sent.
+    /// - `start` indicates whether the operation should start by a START
+    ///   condition and sending the address.
+    /// - `cmd_iterator` is an iterator over the command registers.
     fn setup_write<'a, I>(
         &self,
         addr: u8,
@@ -1503,12 +1526,13 @@ pub trait Instance: crate::private::Sealed {
     }
 
     /// Configures the I2C peripheral for a read operation.
-    /// `addr` is the address of the slave device.
-    /// `buffer` is the buffer to store the read data.
-    /// `start` indicates whether the operation should start by writing the
-    /// address `will_continue` indicates whether there is another read
-    /// operation following this one and we should not nack the last byte.
-    /// The `cmd_iterator` parameter is an iterator over the command registers.
+    /// - `addr` is the address of the slave device.
+    /// - `buffer` is the buffer to store the read data.
+    /// - `start` indicates whether the operation should start by a START
+    ///   condition and sending the address.
+    /// - `will_continue` indicates whether there is another read operation
+    ///   following this one and we should not nack the last byte.
+    /// - `cmd_iterator` is an iterator over the command registers.
     fn setup_read<'a, I>(
         &self,
         addr: u8,
@@ -1915,6 +1939,13 @@ pub trait Instance: crate::private::Sealed {
     }
 
     /// Executes an I2C write operation.
+    /// - `addr` is the address of the slave device.
+    /// - `bytes` is the data two be sent.
+    /// - `start` indicates whether the operation should start by a START
+    ///   condition and sending the address.
+    /// - `stop` indicates whether the operation should end with a STOP
+    ///   condition.
+    /// - `cmd_iterator` is an iterator over the command registers.
     fn write_operation<'a, I>(
         &self,
         address: u8,
@@ -1948,6 +1979,15 @@ pub trait Instance: crate::private::Sealed {
     }
 
     /// Executes an I2C read operation.
+    /// - `addr` is the address of the slave device.
+    /// - `buffer` is the buffer to store the read data.
+    /// - `start` indicates whether the operation should start by a START
+    ///   condition and sending the address.
+    /// - `stop` indicates whether the operation should end with a STOP
+    ///   condition.
+    /// - `will_continue` indicates whether there is another read operation
+    ///   following this one and we should not nack the last byte.
+    /// - `cmd_iterator` is an iterator over the command registers.
     fn read_operation<'a, I>(
         &self,
         address: u8,

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -769,9 +769,9 @@ mod asynch {
         where
             I: Iterator<Item = &'a COMD>,
         {
-            // Short circuit for zero length writes without start as that would be an
+            // Short circuit for zero length writes without start or end as that would be an
             // invalid operation write lengths in the TRM (at least for ESP32-S3) are 1-255
-            if bytes.is_empty() && !start {
+            if bytes.is_empty() && !start && !stop {
                 return Ok(());
             }
 
@@ -1516,10 +1516,6 @@ pub trait Instance: crate::private::Sealed {
     where
         I: Iterator<Item = &'a COMD>,
     {
-        if bytes.is_empty() && !start {
-            return Err(Error::InvalidZeroLength);
-        }
-
         // if start is true we can only send 254 additional bytes with the address as
         // the first
         let max_len = if start { 254usize } else { 255usize };
@@ -1528,16 +1524,19 @@ pub trait Instance: crate::private::Sealed {
             return Err(Error::ExceedingFifo);
         }
 
-        let extra_len = if start { 1 } else { 0 };
-        // WRITE command
-        add_cmd(
-            cmd_iterator,
-            Command::Write {
-                ack_exp: Ack::Ack,
-                ack_check_en: true,
-                length: extra_len + bytes.len() as u8,
-            },
-        )?;
+        let write_len = if start { bytes.len() + 1 } else { bytes.len() };
+        // don't issue write if there is no data to write
+        if write_len > 0 {
+            // WRITE command
+            add_cmd(
+                cmd_iterator,
+                Command::Write {
+                    ack_exp: Ack::Ack,
+                    ack_check_en: true,
+                    length: write_len as u8,
+                },
+            )?;
+        }
 
         self.update_config();
 
@@ -1986,9 +1985,9 @@ pub trait Instance: crate::private::Sealed {
     where
         I: Iterator<Item = &'a COMD>,
     {
-        // Short circuit for zero length writes without start as that would be an
+        // Short circuit for zero length writes without start or end as that would be an
         // invalid operation write lengths in the TRM (at least for ESP32-S3) are 1-255
-        if bytes.is_empty() && !start {
+        if bytes.is_empty() && !start && !stop {
             return Ok(());
         }
 

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -106,6 +106,17 @@ enum Op {
     None,
 }
 
+impl From<Option<&&mut embedded_hal::i2c::Operation<'_>>> for Op {
+    fn from(op: Option<&&mut embedded_hal::i2c::Operation<'_>>) -> Self {
+        use embedded_hal::i2c::Operation;
+        match op {
+            Some(Operation::Write(_)) => Op::Write,
+            Some(Operation::Read(_)) => Op::Read,
+            None => Op::None,
+        }
+    }
+}
+
 impl embedded_hal::i2c::Error for Error {
     fn kind(&self) -> embedded_hal::i2c::ErrorKind {
         use embedded_hal::i2c::{ErrorKind, NoAcknowledgeSource};
@@ -257,16 +268,7 @@ where
         let mut last_op = Op::None;
         let mut op_iter = operations.iter_mut().peekable();
         while let Some(op) = op_iter.next() {
-            let next_op = {
-                if let Some(next_op) = op_iter.peek() {
-                    match next_op {
-                        Operation::Write(_) => Op::Write,
-                        Operation::Read(_) => Op::Read,
-                    }
-                } else {
-                    Op::None
-                }
-            };
+            let next_op: Op = op_iter.peek().into();
             // Clear all I2C interrupts
             self.peripheral.clear_all_interrupts();
 
@@ -927,16 +929,7 @@ mod asynch {
             let mut last_op = Op::None;
             let mut op_iter = operations.iter_mut().peekable();
             while let Some(op) = op_iter.next() {
-                let next_op = {
-                    if let Some(next_op) = op_iter.peek() {
-                        match next_op {
-                            Operation::Write(_) => Op::Write,
-                            Operation::Read(_) => Op::Read,
-                        }
-                    } else {
-                        Op::None
-                    }
-                };
+                let next_op: Op = op_iter.peek().into();
                 // Clear all I2C interrupts
                 self.peripheral.clear_all_interrupts();
 

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -1540,7 +1540,12 @@ pub trait Instance: crate::private::Sealed {
         let (max_len, initial_len) = if will_continue {
             (255usize, buffer.len())
         } else {
-            (254usize, buffer.len() - 1)
+            let len = if buffer.is_empty() {
+                0
+            } else {
+                buffer.len() - 1
+            };
+            (254usize, len)
         };
         if buffer.len() > max_len {
             // we could support more by adding multiple read operations
@@ -1570,7 +1575,9 @@ pub trait Instance: crate::private::Sealed {
             )?;
         }
 
-        if !will_continue {
+        // If we are not continuing the read in the next operation we need to nack the
+        // last byte. But don't issue the read command if the buffer is empty!
+        if !will_continue && !buffer.is_empty() {
             // this is the last read so we need to nack the last byte
             // READ w/o ACK
             add_cmd(

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -769,6 +769,12 @@ mod asynch {
         where
             I: Iterator<Item = &'a COMD>,
         {
+            // Short circuit for zero length writes without start as that would be an
+            // invalid operation write lengths in the TRM (at least for ESP32-S3) are 1-255
+            if bytes.is_empty() && !start {
+                return Ok(());
+            }
+
             // Reset FIFO and command list
             self.peripheral.reset_fifo();
             self.peripheral.reset_command_list();
@@ -812,6 +818,12 @@ mod asynch {
         where
             I: Iterator<Item = &'a COMD>,
         {
+            // Short circuit for zero length reads as that would be an invalid operation
+            // read lengths in the TRM (at least for ESP32-S3) are 1-255
+            if buffer.is_empty() {
+                return Ok(());
+            }
+
             // Reset FIFO and command list
             self.peripheral.reset_fifo();
             self.peripheral.reset_command_list();
@@ -1507,7 +1519,9 @@ pub trait Instance: crate::private::Sealed {
         if bytes.is_empty() && !start {
             return Err(Error::InvalidZeroLength);
         }
-        // if start we can only send 254 additional bytes, the address would be the first
+
+        // if start is true we can only send 254 additional bytes with the address as
+        // the first
         let max_len = if start { 254usize } else { 255usize };
         if bytes.len() > max_len {
             // we could support more by adding multiple write operations
@@ -1972,6 +1986,12 @@ pub trait Instance: crate::private::Sealed {
     where
         I: Iterator<Item = &'a COMD>,
     {
+        // Short circuit for zero length writes without start as that would be an
+        // invalid operation write lengths in the TRM (at least for ESP32-S3) are 1-255
+        if bytes.is_empty() && !start {
+            return Ok(());
+        }
+
         // Reset FIFO and command list
         self.reset_fifo();
         self.reset_command_list();
@@ -2015,6 +2035,12 @@ pub trait Instance: crate::private::Sealed {
     where
         I: Iterator<Item = &'a COMD>,
     {
+        // Short circuit for zero length reads as that would be an invalid operation
+        // read lengths in the TRM (at least for ESP32-S3) are 1-255
+        if buffer.is_empty() {
+            return Ok(());
+        }
+
         // Reset FIFO and command list
         self.reset_fifo();
         self.reset_command_list();

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -1477,7 +1477,7 @@ pub trait Instance: crate::private::Sealed {
             return Err(Error::ExceedingFifo);
         }
 
-        let extra_len = if start {1} else {0};
+        let extra_len = if start { 1 } else { 0 };
         // WRITE command
         add_cmd(
             cmd_iterator,

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -92,6 +92,8 @@ pub enum Error {
     ExecIncomplete,
     /// The number of commands issued exceeded the limit.
     CommandNrExceeded,
+    /// Zero length read or write operation.
+    InvalidZeroLength,
 }
 
 #[derive(PartialEq)]
@@ -1489,6 +1491,9 @@ pub trait Instance: crate::private::Sealed {
     where
         I: Iterator<Item = &'a COMD>,
     {
+        if bytes.is_empty() && !start {
+            return Err(Error::InvalidZeroLength);
+        }
         let max_len = if start { 254usize } else { 255usize };
         if bytes.len() > max_len {
             // we could support more by adding multiple write operations
@@ -1537,15 +1542,13 @@ pub trait Instance: crate::private::Sealed {
     where
         I: Iterator<Item = &'a COMD>,
     {
+        if buffer.is_empty() {
+            return Err(Error::InvalidZeroLength);
+        }
         let (max_len, initial_len) = if will_continue {
             (255usize, buffer.len())
         } else {
-            let len = if buffer.is_empty() {
-                0
-            } else {
-                buffer.len() - 1
-            };
-            (254usize, len)
+            (254usize, buffer.len() - 1)
         };
         if buffer.len() > max_len {
             // we could support more by adding multiple read operations
@@ -1575,9 +1578,7 @@ pub trait Instance: crate::private::Sealed {
             )?;
         }
 
-        // If we are not continuing the read in the next operation we need to nack the
-        // last byte. But don't issue the read command if the buffer is empty!
-        if !will_continue && !buffer.is_empty() {
+        if !will_continue {
             // this is the last read so we need to nack the last byte
             // READ w/o ACK
             add_cmd(

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -95,10 +95,11 @@ pub enum Error {
 }
 
 #[derive(PartialEq)]
-// This enum is used to keep track of the last operation that was performed
-// in an embedded-hal(-async) I2C::transaction. It used to determine whether
-// a START condition should be issued at the start of the current operation.
-enum OpWasIs {
+// This enum is used to keep track of the last/next operation that was/will be
+// performed in an embedded-hal(-async) I2C::transaction. It used to determine
+// whether a START condition should be issued at the start of the current
+// operation and whether a read needs an ack or a nack for the final byte.
+enum Op {
     Write,
     Read,
     None,
@@ -252,17 +253,17 @@ where
         operations: &mut [embedded_hal::i2c::Operation<'_>],
     ) -> Result<(), Self::Error> {
         use embedded_hal::i2c::Operation;
-        let mut last_op = OpWasIs::None;
+        let mut last_op = Op::None;
         let mut op_iter = operations.iter_mut().peekable();
         while let Some(op) = op_iter.next() {
             let next_op = {
                 if let Some(next_op) = op_iter.peek() {
                     match next_op {
-                        Operation::Write(_) => OpWasIs::Write,
-                        Operation::Read(_) => OpWasIs::Read,
+                        Operation::Write(_) => Op::Write,
+                        Operation::Read(_) => Op::Read,
                     }
                 } else {
-                    OpWasIs::None
+                    Op::None
                 }
             };
             // Clear all I2C interrupts
@@ -278,11 +279,11 @@ where
                     self.peripheral.write_operation(
                         address,
                         bytes,
-                        last_op != OpWasIs::Write,
+                        last_op != Op::Write,
                         op_iter.peek().is_none(),
                         cmd_iterator,
                     )?;
-                    last_op = OpWasIs::Write;
+                    last_op = Op::Write;
                 }
                 Operation::Read(buffer) => {
                     // execute a read operation:
@@ -292,12 +293,12 @@ where
                     self.peripheral.read_operation(
                         address,
                         buffer,
-                        last_op != OpWasIs::Read,
+                        last_op != Op::Read,
                         op_iter.peek().is_none(),
-                        next_op == OpWasIs::Read,
+                        next_op == Op::Read,
                         cmd_iterator,
                     )?;
-                    last_op = OpWasIs::Read;
+                    last_op = Op::Read;
                 }
             }
         }
@@ -904,17 +905,17 @@ mod asynch {
             address: u8,
             operations: &mut [Operation<'_>],
         ) -> Result<(), Self::Error> {
-            let mut last_op = OpWasIs::None;
+            let mut last_op = Op::None;
             let mut op_iter = operations.iter_mut().peekable();
             while let Some(op) = op_iter.next() {
                 let next_op = {
                     if let Some(next_op) = op_iter.peek() {
                         match next_op {
-                            Operation::Write(_) => OpWasIs::Write,
-                            Operation::Read(_) => OpWasIs::Read,
+                            Operation::Write(_) => Op::Write,
+                            Operation::Read(_) => Op::Read,
                         }
                     } else {
-                        OpWasIs::None
+                        Op::None
                     }
                 };
                 // Clear all I2C interrupts
@@ -926,12 +927,12 @@ mod asynch {
                         self.write_operation(
                             address,
                             bytes,
-                            last_op != OpWasIs::Write,
+                            last_op != Op::Write,
                             op_iter.peek().is_none(),
                             cmd_iterator,
                         )
                         .await?;
-                        last_op = OpWasIs::Write;
+                        last_op = Op::Write;
                     }
                     Operation::Read(buffer) => {
                         // execute a read operation:
@@ -940,13 +941,13 @@ mod asynch {
                         self.read_operation(
                             address,
                             buffer,
-                            last_op != OpWasIs::Read,
+                            last_op != Op::Read,
                             op_iter.peek().is_none(),
-                            next_op == OpWasIs::Read,
+                            next_op == Op::Read,
                             cmd_iterator,
                         )
                         .await?;
-                        last_op = OpWasIs::Read;
+                        last_op = Op::Read;
                     }
                 }
             }

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -96,9 +96,10 @@ pub enum Error {
 
 #[derive(PartialEq)]
 // This enum is used to keep track of the last/next operation that was/will be
-// performed in an embedded-hal(-async) I2C::transaction. It used to determine
-// whether a START condition should be issued at the start of the current
-// operation and whether a read needs an ack or a nack for the final byte.
+// performed in an embedded-hal(-async) I2C::transaction. It is used to
+// determine whether a START condition should be issued at the start of the
+// current operation and whether a read needs an ack or a nack for the final
+// byte.
 enum Op {
     Write,
     Read,

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -268,13 +268,19 @@ where
     ) -> Result<(), Self::Error> {
         use embedded_hal::i2c::Operation;
         let mut last_op = Op::None;
-        let mut op_iter = operations.iter_mut().peekable();
+        // filter out 0 length read operations
+        let mut op_iter = operations
+            .iter_mut()
+            .filter(|op| match op {
+                Operation::Write(_) => true,
+                Operation::Read(buffer) => !buffer.is_empty(),
+            })
+            .peekable();
         while let Some(op) = op_iter.next() {
             let next_op: Op = op_iter.peek().into();
             // Clear all I2C interrupts
             self.peripheral.clear_all_interrupts();
 
-            // TODO somehow know that we can combine a write and a read into one transaction
             let cmd_iterator = &mut self.peripheral.register_block().comd_iter();
             match op {
                 Operation::Write(bytes) => {
@@ -929,7 +935,14 @@ mod asynch {
             operations: &mut [Operation<'_>],
         ) -> Result<(), Self::Error> {
             let mut last_op = Op::None;
-            let mut op_iter = operations.iter_mut().peekable();
+            // filter out 0 length read operations
+            let mut op_iter = operations
+                .iter_mut()
+                .filter(|op| match op {
+                    Operation::Write(_) => true,
+                    Operation::Read(buffer) => !buffer.is_empty(),
+                })
+                .peekable();
             while let Some(op) = op_iter.next() {
                 let next_op: Op = op_iter.peek().into();
                 // Clear all I2C interrupts

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -895,11 +895,13 @@ mod asynch {
                 addr,
                 bytes,
                 true,
-                false,
+                buffer.is_empty(), // if the read buffer is empty, then issue a stop
                 &mut self.peripheral.register_block().comd_iter(),
             )
             .await?;
             self.peripheral.clear_all_interrupts();
+            // this will be a no-op if the buffer is empty, in that case we issued the stop
+            // with the write
             self.read_operation(
                 addr,
                 buffer,
@@ -2111,10 +2113,12 @@ pub trait Instance: crate::private::Sealed {
             addr,
             bytes,
             true,
-            false,
+            buffer.is_empty(), // if the read buffer is empty, then issue a stop
             &mut self.register_block().comd_iter(),
         )?;
         self.clear_all_interrupts();
+        // this will be a no-op if the buffer is empty, in that case we issued the stop
+        // with the write
         self.read_operation(
             addr,
             buffer,

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -98,7 +98,7 @@ pub enum Error {
 // This enum is used to keep track of the last operation that was performed
 // in an embedded-hal(-async) I2C::transaction. It used to determine whether
 // a START condition should be issued at the start of the current operation.
-enum LastOpWas {
+enum OpWasIs {
     Write,
     Read,
     None,
@@ -252,9 +252,19 @@ where
         operations: &mut [embedded_hal::i2c::Operation<'_>],
     ) -> Result<(), Self::Error> {
         use embedded_hal::i2c::Operation;
-        let mut last_op = LastOpWas::None;
+        let mut last_op = OpWasIs::None;
         let mut op_iter = operations.iter_mut().peekable();
         while let Some(op) = op_iter.next() {
+            let next_op = {
+                if let Some(next_op) = op_iter.peek() {
+                    match next_op {
+                        Operation::Write(_) => OpWasIs::Write,
+                        Operation::Read(_) => OpWasIs::Read,
+                    }
+                } else {
+                    OpWasIs::None
+                }
+            };
             // Clear all I2C interrupts
             self.peripheral.clear_all_interrupts();
 
@@ -268,24 +278,26 @@ where
                     self.peripheral.write_operation(
                         address,
                         bytes,
-                        last_op != LastOpWas::Write,
+                        last_op != OpWasIs::Write,
                         op_iter.peek().is_none(),
                         cmd_iterator,
                     )?;
-                    last_op = LastOpWas::Write;
+                    last_op = OpWasIs::Write;
                 }
                 Operation::Read(buffer) => {
                     // execute a read operation:
                     // - issue START/RSTART if op is different from previous
                     // - issue STOP if op is the last one
+                    // - will_continue is true if there is another read operation next
                     self.peripheral.read_operation(
                         address,
                         buffer,
-                        last_op != LastOpWas::Read,
+                        last_op != OpWasIs::Read,
                         op_iter.peek().is_none(),
+                        next_op == OpWasIs::Read,
                         cmd_iterator,
                     )?;
-                    last_op = LastOpWas::Read;
+                    last_op = OpWasIs::Read;
                 }
             }
         }
@@ -743,7 +755,8 @@ mod asynch {
             if start {
                 add_cmd(cmd_iterator, Command::Start)?;
             }
-            self.peripheral.setup_write(address, bytes, cmd_iterator)?;
+            self.peripheral
+                .setup_write(address, bytes, start, cmd_iterator)?;
             add_cmd(
                 cmd_iterator,
                 if stop { Command::Stop } else { Command::End },
@@ -763,6 +776,7 @@ mod asynch {
             buffer: &mut [u8],
             start: bool,
             stop: bool,
+            will_continue: bool,
             cmd_iterator: &mut I,
         ) -> Result<(), Error>
         where
@@ -774,7 +788,8 @@ mod asynch {
             if start {
                 add_cmd(cmd_iterator, Command::Start)?;
             }
-            self.peripheral.setup_read(address, buffer, cmd_iterator)?;
+            self.peripheral
+                .setup_read(address, buffer, start, will_continue, cmd_iterator)?;
             add_cmd(
                 cmd_iterator,
                 if stop { Command::Stop } else { Command::End },
@@ -812,6 +827,7 @@ mod asynch {
                 buffer,
                 true,
                 true,
+                false,
                 &mut self.peripheral.register_block().comd_iter(),
             )
             .await?;
@@ -847,6 +863,7 @@ mod asynch {
                 buffer,
                 true,
                 true,
+                false,
                 &mut self.peripheral.register_block().comd_iter(),
             )
             .await?;
@@ -887,9 +904,19 @@ mod asynch {
             address: u8,
             operations: &mut [Operation<'_>],
         ) -> Result<(), Self::Error> {
-            let mut last_op = LastOpWas::None;
+            let mut last_op = OpWasIs::None;
             let mut op_iter = operations.iter_mut().peekable();
             while let Some(op) = op_iter.next() {
+                let next_op = {
+                    if let Some(next_op) = op_iter.peek() {
+                        match next_op {
+                            Operation::Write(_) => OpWasIs::Write,
+                            Operation::Read(_) => OpWasIs::Read,
+                        }
+                    } else {
+                        OpWasIs::None
+                    }
+                };
                 // Clear all I2C interrupts
                 self.peripheral.clear_all_interrupts();
 
@@ -899,12 +926,12 @@ mod asynch {
                         self.write_operation(
                             address,
                             bytes,
-                            last_op != LastOpWas::Write,
+                            last_op != OpWasIs::Write,
                             op_iter.peek().is_none(),
                             cmd_iterator,
                         )
                         .await?;
-                        last_op = LastOpWas::Write;
+                        last_op = OpWasIs::Write;
                     }
                     Operation::Read(buffer) => {
                         // execute a read operation:
@@ -913,12 +940,13 @@ mod asynch {
                         self.read_operation(
                             address,
                             buffer,
-                            last_op != LastOpWas::Read,
+                            last_op != OpWasIs::Read,
                             op_iter.peek().is_none(),
+                            next_op == OpWasIs::Read,
                             cmd_iterator,
                         )
                         .await?;
-                        last_op = LastOpWas::Read;
+                        last_op = OpWasIs::Read;
                     }
                 }
             }
@@ -1433,86 +1461,111 @@ pub trait Instance: crate::private::Sealed {
     }
 
     /// Configures the I2C peripheral for a write operation.
-    fn setup_write<'a, I>(&self, addr: u8, bytes: &[u8], cmd_iterator: &mut I) -> Result<(), Error>
-    where
-        I: Iterator<Item = &'a COMD>,
-    {
-        if bytes.len() > 254 {
-            // we could support more by adding multiple write operations
-            return Err(Error::ExceedingFifo);
-        }
-
-        // WRITE command
-        add_cmd(
-            cmd_iterator,
-            Command::Write {
-                ack_exp: Ack::Ack,
-                ack_check_en: true,
-                length: 1 + bytes.len() as u8,
-            },
-        )?;
-
-        self.update_config();
-
-        // Load address and R/W bit into FIFO
-        write_fifo(
-            self.register_block(),
-            addr << 1 | OperationType::Write as u8,
-        );
-
-        Ok(())
-    }
-
-    /// Configures the I2C peripheral for a read operation.
-    fn setup_read<'a, I>(
+    fn setup_write<'a, I>(
         &self,
         addr: u8,
-        buffer: &mut [u8],
+        bytes: &[u8],
+        start: bool,
         cmd_iterator: &mut I,
     ) -> Result<(), Error>
     where
         I: Iterator<Item = &'a COMD>,
     {
-        if buffer.len() > 254 {
-            // we could support more by adding multiple read operations
+        let max_len = if start { 254usize } else { 255usize };
+        if bytes.len() > max_len {
+            // we could support more by adding multiple write operations
             return Err(Error::ExceedingFifo);
         }
 
+        let extra_len = if start {1} else {0};
         // WRITE command
         add_cmd(
             cmd_iterator,
             Command::Write {
                 ack_exp: Ack::Ack,
                 ack_check_en: true,
-                length: 1,
-            },
-        )?;
-
-        if buffer.len() > 1 {
-            // READ command (N - 1)
-            add_cmd(
-                cmd_iterator,
-                Command::Read {
-                    ack_value: Ack::Ack,
-                    length: buffer.len() as u8 - 1,
-                },
-            )?;
-        }
-
-        // READ w/o ACK
-        add_cmd(
-            cmd_iterator,
-            Command::Read {
-                ack_value: Ack::Nack,
-                length: 1,
+                length: extra_len + bytes.len() as u8,
             },
         )?;
 
         self.update_config();
 
-        // Load address and R/W bit into FIFO
-        write_fifo(self.register_block(), addr << 1 | OperationType::Read as u8);
+        if start {
+            // Load address and R/W bit into FIFO
+            write_fifo(
+                self.register_block(),
+                addr << 1 | OperationType::Write as u8,
+            );
+        }
+        Ok(())
+    }
 
+    /// Configures the I2C peripheral for a read operation.
+    /// `addr` is the address of the slave device.
+    /// `buffer` is the buffer to store the read data.
+    /// `start` indicates whether the operation should start by writing the
+    /// address `will_continue` indicates whether there is another read
+    /// operation following this one and we should not nack the last byte.
+    /// The `cmd_iterator` parameter is an iterator over the command registers.
+    fn setup_read<'a, I>(
+        &self,
+        addr: u8,
+        buffer: &mut [u8],
+        start: bool,
+        will_continue: bool,
+        cmd_iterator: &mut I,
+    ) -> Result<(), Error>
+    where
+        I: Iterator<Item = &'a COMD>,
+    {
+        let (max_len, initial_len) = if will_continue {
+            (255usize, buffer.len())
+        } else {
+            (254usize, buffer.len() - 1)
+        };
+        if buffer.len() > max_len {
+            // we could support more by adding multiple read operations
+            return Err(Error::ExceedingFifo);
+        }
+
+        if start {
+            // WRITE command
+            add_cmd(
+                cmd_iterator,
+                Command::Write {
+                    ack_exp: Ack::Ack,
+                    ack_check_en: true,
+                    length: 1,
+                },
+            )?;
+        }
+        // there is another read so we need to ack all
+        add_cmd(
+            cmd_iterator,
+            Command::Read {
+                ack_value: Ack::Ack,
+                length: initial_len as u8,
+            },
+        )?;
+
+        if !will_continue {
+            // this is the last read so we need to nack the last byte
+            // READ w/o ACK
+            add_cmd(
+                cmd_iterator,
+                Command::Read {
+                    ack_value: Ack::Nack,
+                    length: 1,
+                },
+            )?;
+        }
+
+        self.update_config();
+
+        if start {
+            // Load address and R/W bit into FIFO
+            write_fifo(self.register_block(), addr << 1 | OperationType::Read as u8);
+        }
         Ok(())
     }
 
@@ -1875,7 +1928,7 @@ pub trait Instance: crate::private::Sealed {
         if start {
             add_cmd(cmd_iterator, Command::Start)?;
         }
-        self.setup_write(address, bytes, cmd_iterator)?;
+        self.setup_write(address, bytes, start, cmd_iterator)?;
         add_cmd(
             cmd_iterator,
             if stop { Command::Stop } else { Command::End },
@@ -1896,6 +1949,7 @@ pub trait Instance: crate::private::Sealed {
         buffer: &mut [u8],
         start: bool,
         stop: bool,
+        will_continue: bool,
         cmd_iterator: &mut I,
     ) -> Result<(), Error>
     where
@@ -1908,7 +1962,9 @@ pub trait Instance: crate::private::Sealed {
         if start {
             add_cmd(cmd_iterator, Command::Start)?;
         }
-        self.setup_read(address, buffer, cmd_iterator)?;
+
+        self.setup_read(address, buffer, start, will_continue, cmd_iterator)?;
+
         add_cmd(
             cmd_iterator,
             if stop { Command::Stop } else { Command::End },
@@ -1945,6 +2001,7 @@ pub trait Instance: crate::private::Sealed {
             buffer,
             true,
             true,
+            false,
             &mut self.register_block().comd_iter(),
         )?;
         Ok(())
@@ -1978,6 +2035,7 @@ pub trait Instance: crate::private::Sealed {
             buffer,
             true,
             true,
+            false,
             &mut self.register_block().comd_iter(),
         )?;
         Ok(())

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -1494,6 +1494,7 @@ pub trait Instance: crate::private::Sealed {
         if bytes.is_empty() && !start {
             return Err(Error::InvalidZeroLength);
         }
+        // if start we can only send 254 additional bytes, the address would be the first
         let max_len = if start { 254usize } else { 255usize };
         if bytes.len() > max_len {
             // we could support more by adding multiple write operations

--- a/esp-hal/src/i2c.rs
+++ b/esp-hal/src/i2c.rs
@@ -1540,14 +1540,17 @@ pub trait Instance: crate::private::Sealed {
                 },
             )?;
         }
-        // there is another read so we need to ack all
-        add_cmd(
-            cmd_iterator,
-            Command::Read {
-                ack_value: Ack::Ack,
-                length: initial_len as u8,
-            },
-        )?;
+
+        if initial_len > 0 {
+            // READ command
+            add_cmd(
+                cmd_iterator,
+                Command::Read {
+                    ack_value: Ack::Ack,
+                    length: initial_len as u8,
+                },
+            )?;
+        }
 
         if !will_continue {
             // this is the last read so we need to nack the last byte

